Add wrapper for `get_formatted_meta_data` to make `$include_all` default to `true`

opr opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR proposes the introduction of a new method called `get_all_formatted_meta_data` in the `WC_Order_Item` class. The purpose of this is to allow us to call `get_formatted_meta_data` with `$include_all = true` while leaving the existing functionality of `get_formatted_meta_data` intact for back-compatibility with whoever might be using it.

The PR also changes uses of `get_formatted_meta_data` in _this_ repository to use the new method.

The current behaviour _without_ this change is that metadata line items are skipped if the _value_ is the same as one that is previously shown.

In some cases, this is problematic, for example if using WooCommerce Product Addons on a variable product and one of the addons has the same value as the selected attribute.

Without this change, the below image would only show the size, and not the trophy cup inscription.

<img src="https://user-images.githubusercontent.com/5656702/137300211-9f62f6b2-77a7-4186-b7f4-07a774b40431.png" width=400 />

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes 645-gh-woocommerce/woocommerce-product-addons.
### How to test the changes in this Pull Request:

1. Create a variable product with attribute Size: Small, Medium, Large, Extra Large
2. Check out the PR in Product Addons: 705-gh-woocommerce/woocommerce-product-addons
3. Create a global add-on that has these 4 options + some extra ones (you can import these as a reference):

```
a:1:{i:0;a:16:{s:4:"name";s:22:"Trophy Cup Inscription";s:12:"title_format";s:5:"label";s:18:"description_enable";i:1;s:11:"description";s:101:"Please select the option that applies to your cup size if you require your trophy cup to be engraved.";s:4:"type";s:15:"multiple_choice";s:7:"display";s:6:"select";s:8:"position";i:0;s:8:"required";i:1;s:12:"restrictions";i:0;s:17:"restrictions_type";s:8:"any_text";s:12:"adjust_price";i:0;s:10:"price_type";s:8:"flat_fee";s:5:"price";s:0:"";s:3:"min";d:0;s:3:"max";d:0;s:7:"options";a:13:{i:0;a:4:{s:5:"label";s:8:"4X Small";s:5:"price";s:5:"20.00";s:5:"image";s:0:"";s:10:"price_type";s:8:"flat_fee";}i:1;a:4:{s:5:"label";s:8:"3X Small";s:5:"price";s:5:"20.00";s:5:"image";s:0:"";s:10:"price_type";s:8:"flat_fee";}i:2;a:4:{s:5:"label";s:8:"2X Small";s:5:"price";s:5:"20.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:3;a:4:{s:5:"label";s:7:"X Small";s:5:"price";s:5:"25.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:4;a:4:{s:5:"label";s:5:"Small";s:5:"price";s:5:"30.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:5;a:4:{s:5:"label";s:6:"Medium";s:5:"price";s:5:"35.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:6;a:4:{s:5:"label";s:5:"Large";s:5:"price";s:5:"40.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:7;a:4:{s:5:"label";s:7:"X Large";s:5:"price";s:5:"45.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:8;a:4:{s:5:"label";s:8:"2X Large";s:5:"price";s:5:"50.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:9;a:4:{s:5:"label";s:8:"3X Large";s:5:"price";s:5:"55.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:10;a:4:{s:5:"label";s:8:"4X large";s:5:"price";s:5:"60.00";s:5:"image";s:0:"";s:10:"price_type";s:8:"flat_fee";}i:11;a:4:{s:5:"label";s:22:"Engraved Plate on Base";s:5:"price";s:5:"15.00";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}i:12;a:4:{s:5:"label";s:12:"No engraving";s:5:"price";s:0:"";s:5:"image";s:0:"";s:10:"price_type";s:14:"quantity_based";}}}}
```
4. Add a simple product to your cart.
5. Add the variable product from step 1 with the variation `Size: Large` and add-on option `Large` (which appears in the attribute list) and complete the order.
6. Go to the order in the WooCommerce dashboard and ensure the attribute Size: Large and the Addon with the value Large are both displaying.
7. Ensure the simple product's addon value is displaying too.
7. Delete the global addon and check out normally with a **variable** product. Go to the order in the WooCommerce dashboard and ensure the variations are shown as expected.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable? **I don't think this requires tests, but I can't run tests locally - see p1634213755141500-slack-C8X6Q7XQU** 
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Add 'woocommerce_get_formatted_meta_data_include_all_meta_lines' filter hook. This can be used to control whether metadata lines are shown in the order meta box.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

opr committed:
Add 'woocommerce_show_product_variant_metadata_line' hook

This can be used to allow extensions to override whether the metadata line
is displayed in the order meta box or not.

nerrad commented on includes/class-wc-order-item.php:
```diff
@@ -279,8 +279,19 @@ public function get_formatted_meta_data( $hideprefix = '_', $include_all = false
+				$display_value,
+				$product,
+				$show_metadata_line
+			);
```
`$show_metadata_line` should be the first attribute returned here, otherwise you're changing the original behaviour.

nerrad reviewed (commented):
I think I understand the need for these changes, but one problem with the current implementation (ignoring the obvious incorrect filter implementation) is that it seems far too easy for extensions to ignore the explicit behaviour requested by the usage of the `get_formatted_meta_data` function (particularly the usage of `$include_all`). While you could rectify this by ensuring `$include_all` is passed along to the filter, it still introduces a higher risk of breakage imo.

Offhand, I'm not sure what a better solution for the fix you are trying to make would be. Have you investigated whether there are any other options yet?

opr committed:
Add woocommerce_order_item_meta_include_all_meta_lines filter

This will be used to determine if all meta lines should be shown in the order-item-meta template.

opr committed:
Update uses of get_formatted_metadata to pass through filter first

opr committed:
Remove accidentally added empty line

opr committed:
Add file meta to html-order-item-meta.php

opr commented:
> I think I understand the need for these changes, but one problem with the current implementation (ignoring the obvious incorrect filter implementation) is that it seems far too easy for extensions to ignore the explicit behaviour requested by the usage of the `get_formatted_meta_data` function (particularly the usage of `$include_all`). While you could rectify this by ensuring `$include_all` is passed along to the filter, it still introduces a higher risk of breakage imo.
> 
> Offhand, I'm not sure what a better solution for the fix you are trying to make would be. Have you investigated whether there are any other options yet?

Thanks for the review @nerrad I took what you suggested in Slack and investigated if we could put something in place earlier. We can modify the value of `$include_all` higher up the stack instead. However it turns out there was no existing hooks/filters to do this, so I've added a filter.

I've updated the PR description to give more information about where this was changed.

nerrad commented:
Another option...

I wonder if we should be explicit and introduce a new filtered value to the condition here? Something like:

```php
if ( 
  ( ! $include_all || ! apply_filters( 'woocommerce_get_formatted_meta_data_include_all_meta_lines', false, $this->item ) )
  && $product 
  && $product->is_type( 'variation' ) 
  && wc_is_attribute_in_product_name( $display_value, $order_item_name )
) { /** ... **/ }
```

(btw there's a good opportunity to improve the `$product` condition here by doing an `instanceof` check instead of just the current truthy check which doesn't necessarily guarantee it's product)

vedanshujain commented:
I would say `false` is a bad default everywhere except for in `includes/admin/list-tables/class-wc-admin-list-table-orders.php`. Although I am but unsure about `includes/wc-template-functions.php` and need some investigation to confirm. 

I think we can change the  `get_formatted_meta_data` to pass true for `include_all` everywhere we show detailed data, which is pretty much most places, even if the attribute value is part of the item name. If we do that, will it fix the issue with addons?

vedanshujain commented:
@opr Are you still planning to work on this? I think we should just the change the default of `include_all` param to true (or perhaps create another wrapper method to preserve BC).

opr commented:
@vedanshujain yes I'll pick this up next week during Rubik's cooldown 👍🏻

opr committed:
Revert changes made in PR so far

opr committed:
Add docblock to html-order-item-meta

opr committed:
Fix problem with assignment in conditional

opr committed:
Add get_all_formatted_meta_data wrapper

This sets include_all to true

opr committed:
Update uses of get_formatted_data to new wrapper

opr committed:
Add tests for new get_all_formatted_metadata method

opr commented:
@vedanshujain This is now ready to be reviewed - I've updated the main PR description too.

vedanshujain commented on plugins/woocommerce/includes/wc-template-functions.php:
```diff
@@ -3336,7 +3336,7 @@ function wc_display_item_meta( $item, $args = array() ) {
 		);
 
-		foreach ( $item->get_formatted_meta_data() as $meta_id => $meta ) {
+		foreach ( $item->get_all_formatted_meta_data() as $meta_id => $meta ) {
```
IMO we should not display all meta here, since this function is generally used in frontend template. wdyt?

opr replied:
Hey thanks for the review @vedanshujain - I changed this because on the order summary page this is how it displays:
| `get_all_formatted_meta_data` | `get_formatted_meta_data` |
|---|---|
| ![image](https://user-images.githubusercontent.com/5656702/146181483-e20779e1-0681-46ad-99a8-09c1aa0933b2.png) | ![image](https://user-images.githubusercontent.com/5656702/146181535-24edb423-ab7b-4b19-88d3-9385d233e64e.png) |

I think we need to keep it as `get_all_formatted_meta_data` here to solve the issue.

vedanshujain reviewed (commented):
Added a minor comment, otherwise looks good.

vedanshujain reviewed (approved):
I missed notification from this PR, apologies, happy to approve now. Since it's been a lot of time, @opr confirming with you again if we are good to merge.

opr commented:
> I missed notification from this PR, apologies, happy to approve now. Since it's been a lot of time, @opr confirming with you again if we are good to merge.
No problem :) Yes  still looks fine to merge, thanks!

opr commented:
@vedanshujain just pinging to make sure this shows up in your notifications. I just didn't want it to be missed again.

damiannep commented:
@opr We have a merchant complaining that content is now duplicated, for example:

<img width="564" alt="Screen Shot 2022-05-01 at 15 55 57" src="https://user-images.githubusercontent.com/30090682/166149106-2071f6b8-f8da-49c0-8b92-47236a18a1a2.png">

Was this an intentional choice to keep the duplicate meta data?

opr commented:
> Was this an intentional choice to keep the duplicate meta data?

@damiannep sorry for the delay in replying I've been away, I think from my POV, the answer is yes, it is intentional. I created this PR as it fixed a bug in WooCommerce Product Addons, and the discussion above does include screenshots of what will happen after introducing this change. I would seek clarification from the WooCommerce core team though about next steps regarding this.

jnineb commented:
Thanks @opr for the confirmation/update! Will clarify with the WooCommerce core team as well :)

vedanshujain commented:
hi @jnineb yeah this is intentional. the reason is sometimes attributes in the title can be overwritten by other attributes of the same name, extensions, custom code etc. So we decided to enumerate them anyway irrespective of whether they are included in the title or not.
